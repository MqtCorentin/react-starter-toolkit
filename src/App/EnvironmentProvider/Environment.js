import React, { useState, useEffect } from 'react';

const NODE_ENV = process.env.REACT_APP_NODE_ENV || process.env.NODE_ENV;

export const EnvironmentContext = React.createContext();
const { Provider } = EnvironmentContext;

const initState = {
  environment: null,
  error: null,
};

export const getImport = () => import(`../../../public/environment.${NODE_ENV}.json`);

/**
 * Hook to setState environment
 */
export const useEnv = (getImportFn = getImport, initStateCt = initState) => {
  const [envState, setEnvState] = useState(initStateCt);
  useEffect(() => {
    if (envState.environment === null) {
      getImportFn()
        .then(environment => setEnvState({ environment }))
        .catch(error => setEnvState({ error }));
    }
  }, [envState.environment, getImportFn]);

  return { envState, setEnvState };
};

/**
 * Component wrapper to add environment context
 * @param {Component} children
 */
const EnvironmentProvider = ({ children, useEnvFn = useEnv }) => {
  const { envState } = useEnvFn();
  return <Provider value={envState}>{children}</Provider>;
};

export default EnvironmentProvider;
