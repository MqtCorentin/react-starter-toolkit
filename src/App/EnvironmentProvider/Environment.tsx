import { useState, useEffect, createContext, Dispatch, SetStateAction } from 'react';

export type TEnvironment = {
  apiUrl: string;
  baseUrl: string;
  fetchConfig: object;
  oidc: Record<string, boolean | string>;
};

export type TEnvironmentState = {
  environment?: null | TEnvironment;
  error?: unknown;
};

const initState: TEnvironmentState = {
  environment: null,
  error: null,
};

export const EnvironmentContext = createContext<TEnvironmentState>(initState);
EnvironmentContext.displayName = 'EnvironmentContext';

export const getFileEnv = (nodeEnv = process.env.REACT_APP_NODE_ENV || process.env.NODE_ENV) => `environment.${nodeEnv}.json`;

type TfetchEnv = {
  setEnvState: Dispatch<SetStateAction<TEnvironmentState>>;
  signal: AbortSignal;
  fetchFn?: typeof fetch;
  getFileEnvFn?: typeof getFileEnv;
};

export const fetchEnv = async ({ setEnvState, signal, fetchFn = fetch, getFileEnvFn = getFileEnv }: TfetchEnv) => {
  const fileName = getFileEnvFn();
  const response = await fetchFn(`/${fileName}`, { signal }).catch(error => setEnvState({ error }));
  setEnvState({ environment: await response?.json() });
};

/**
 * Hook to setState environment
 */
export const useEnv = (fetchEnvFn = fetchEnv, initStateCt = initState) => {
  const [envState, setEnvState] = useState(initStateCt);
  useEffect(() => {
    const abortController = new AbortController();
    if (envState.environment === null) {
      fetchEnvFn({ setEnvState, signal: abortController.signal });
    }
    return () => abortController.abort();
  }, [envState.environment, fetchEnvFn]);

  return { envState, setEnvState };
};

/**
 * Component wrapper to add environment context
 * @param {Component} children
 */
const EnvironmentProvider = ({ children, useEnvFn = useEnv }: { children: JSX.Element; useEnvFn?: typeof useEnv }) => {
  const { envState } = useEnvFn();
  return <EnvironmentContext.Provider value={envState}>{children}</EnvironmentContext.Provider>;
};

export default EnvironmentProvider;
