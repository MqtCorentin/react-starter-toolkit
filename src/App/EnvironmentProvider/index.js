export { default, EnvironmentContext, useEnv } from './Environment';
