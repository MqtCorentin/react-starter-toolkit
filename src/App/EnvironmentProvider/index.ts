export { default, EnvironmentContext, useEnv } from './Environment';
export type { TEnvironment, TEnvironmentState } from './Environment';
