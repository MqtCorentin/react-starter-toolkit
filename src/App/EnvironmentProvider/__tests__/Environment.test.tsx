import { useContext } from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { render } from '@testing-library/react';
import EnvironmentProvider, { useEnv, EnvironmentContext, fetchEnv } from '..';
import type { TEnvironment } from '..';

const BaseWithEnvironment = ({ environment }: { environment?: TEnvironment | null }) => <div>{environment?.baseUrl}</div>;

const BaseUseEnvContext = () => {
  const envProps = useContext(EnvironmentContext);
  return <BaseWithEnvironment {...envProps} />;
};

const useEnvFnMock = jest.fn().mockReturnValue({
  envState: {
    environment: {
      baseUrl: 'urllocal',
    },
  },
});

const App = () => (
  <EnvironmentProvider useEnvFn={useEnvFnMock}>
    <BaseUseEnvContext />
  </EnvironmentProvider>
);

describe('Render App with Base have env props', () => {
  it('Should render App baseUrl props when call setFetchCustom', () => {
    const { asFragment, getByText } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('urllocal')).toBeDefined();
  });
});

describe('useEnv', () => {
  const fetchEnvFn = jest.fn();
  it('Should import file when state environment is null', () => {
    const initStateCt = {
      environment: null,
    };
    const { result } = renderHook(() => useEnv(fetchEnvFn, initStateCt));

    expect(result.current).toEqual({ envState: { environment: null }, setEnvState: result.current.setEnvState });
    expect(fetchEnvFn).toHaveBeenCalled();
  });
  it('Should not import file config when state environment is not null', () => {
    const environment = {
      apiUrl: '/api',
      baseUrl: '/',
      fetchConfig: {},
      oidc: {},
    };
    const { result } = renderHook(() => useEnv(fetchEnvFn, { environment }));

    expect(result.current).toEqual({ envState: { environment }, setEnvState: result.current.setEnvState });
    expect(fetchEnvFn).not.toHaveBeenCalled();
  });
});

describe('fetchEnv', () => {
  const resolvedValue = { json: () => ({ data: 'data' }), blob: () => ({ blob: 'blob' }), status: 200 };
  const fetchMock = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({ data: 'data' }),
    }),
  ) as jest.Mock;

  const setEnvStateMock = jest.fn();
  const getFileEnvFnMock = jest.fn();

  it('Should getFileEnvFn called and fetchMock called with "/filename" when call fetchEnv', async () => {
    getFileEnvFnMock.mockReturnValue('filename');
    fetchMock.mockResolvedValue(resolvedValue);
    await fetchEnv({ setEnvState: setEnvStateMock, signal: {} as AbortSignal, fetchFn: fetchMock, getFileEnvFn: getFileEnvFnMock });

    expect(getFileEnvFnMock).toBeCalled();
    expect(fetchMock).toBeCalledWith('/filename', { signal: {} });
  });
  it('Should setEnvState called when call fetchEnv', async () => {
    fetchMock.mockResolvedValue(resolvedValue);
    await fetchEnv({ setEnvState: setEnvStateMock, signal: {} as AbortSignal, fetchFn: fetchMock });
    expect(setEnvStateMock).toBeCalled();
  });

  it('Should setEnvState called when call fetchEnv', async () => {
    fetchMock.mockRejectedValue('error');
    try {
      await fetchEnv({ setEnvState: setEnvStateMock, signal: {} as AbortSignal, fetchFn: fetchMock });
    } catch (error) {
      expect(error).toMatch('error');
    }
  });
});

/* describe('useEnv', () => {
  it('Should import file when state environment is null', async () => {
    const getImportMock = jest.fn().mockImplementation(() => Promise.resolve({ client_id: 'test' }));
    const { result } = renderHook(() => useEnv(getImportMock));

    const expectedEnv = { envState: { environment: { client_id: 'test' }, error: null }, setEnvState: result.current.setEnvState };

    await act(async () => {
      expect(result.current).not.toEqual(expectedEnv);
    });
    expect(result.current).toEqual(expectedEnv);
  });

  it('Should not import file config when state environment is not null', () => {
    const environmentMock: TEnvironment = {
      oidc: { client_id: 'interactive.public' },
      fetchConfig: { method: 'GET' },
      apiUrl: 'url',
      baseUrl: 'url',
    };

    const mockEnv: TEnvironmentState = { environment: environmentMock, error: null };
    const { result } = renderHook(() => useEnv(() => Promise.resolve(), mockEnv));
    const expectedEnv = { envState: mockEnv, setEnvState: result.current.setEnvState };

    act(() => {
      result.current.setEnvState({
        environment: environmentMock,
        error: null,
      });
    });

    expect(result.current).toEqual(expectedEnv);
  });
}); */
