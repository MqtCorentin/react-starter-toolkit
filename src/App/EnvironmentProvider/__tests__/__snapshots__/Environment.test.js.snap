// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App with Base have env props Should render App baseUrl props when call setFetchCustom 1`] = `
<DocumentFragment>
  <div>
    urllocal
  </div>
</DocumentFragment>
`;
